crates/typed-store/src/rocks/tests.rs

```diff
@@ -925,6 +928,129 @@ async fn open_as_secondary_test() {
     assert_eq!(secondary_db.get(&0).unwrap(), Some("10".to_string()));
 }
 
+#[derive(Serialize, Deserialize, Copy, Clone)]
+struct ObjectWithRefCount {
+    value: i64,
+    ref_count: i64,
+}
+
+fn increment_counter(db: &DBMap<String, ObjectWithRefCount>, key: &str, value: i64) {
+    let batch = db
+        .batch()
+        .partial_merge_batch(db, [(key.to_string(), value.to_le_bytes())])
+        .unwrap();
+    batch.write().unwrap();
+}
+
+#[tokio::test]
+async fn refcount_test() {
+    let key = "key".to_string();
+    let mut options = rocksdb::Options::default();
+    options.set_merge_operator(
+        "refcount operator",
+        reference_count_merge_operator,
+        reference_count_merge_operator,
+    );
+    let db = DBMap::<String, ObjectWithRefCount>::open(
+        temp_dir(),
+        MetricConf::default(),
+        Some(options),
+        None,
+        &ReadWriteOptions::default(),
+    )
+    .expect("failed to open rocksdb");
+    let object = ObjectWithRefCount {
+        value: 3,
+        ref_count: 1,
+    };
+    // increment value 10 times
+    let iterations = 10;
+    for _ in 0..iterations {
+        db.batch()
+            .merge_batch(&db, [(key.to_string(), object)])
+            .unwrap()
+            .write()
+            .unwrap();
+    }
+    let value = db
+        .get(&key)
+        .expect("failed to read value")
+        .expect("value is empty");
+    assert_eq!(value.value, object.value);
+    assert_eq!(value.ref_count, iterations);
+
+    // decrement value
+    increment_counter(&db, &key, -1);
+    let value = db.get(&key).unwrap().unwrap();
+    assert_eq!(value.value, object.value);
+    assert_eq!(value.ref_count, iterations - 1);
+}
+
+#[tokio::test]
+async fn refcount_with_compaction_test() {
```

Comment:
Let's resume the previous conversation regarding space usage and transactions. It looks like there's a way to achieve desired goals without transactions.

To refresh, the current approach doesn't require transactions, but it's space inefficient. Each new object update performs a merge operation with the full value. Eventually, space is getting reclaimed during the compaction process(redundant merge operators are getting squashed), but it may take a while.

Let's consider a naive approach where during transaction execution we do read before write(non-transactional). And if `get` returns nothing we perform merge with a full value, if `get` returns something, we do a partial merge (i.e. incrementing ref count value by 1).

Without race conditions, we should expect only one full merge(the first one) and a bunch of increments and decrements caused by transaction executions and pruning jobs respectively.   

Now let's consider possible race conditions:
1. Two concurrent updates happening at the same time, perform get before write and observe that value doesn't exist. So they both will initiate a merge operation with a full value. That's ok, all subsequent reads will be able to merge those operations, and the compaction job eventually will squash two full values and reclaim the space
2. The race between the update and the pruner. The sequence of troublesome steps:
   * Transaction execution checks that value already exists
   * Pruner decrements ref count value and it's 0 now
   * Compaction job merges all operands to an empty value (because ref count is 0 now)
   * Transaction execution issues ref count increment instead of a merge with a full value
 
Now data is corrupted and we can't read the original value. Such a race seems to be very unlikely, but let's say it's possible. 

The solution is to issue rocksdb's snapshot operation somewhere before performing decrement. The compaction process respects the snapshot's supporting operations and doesn't squash merge operands if they still might be in use.
`refcount_with_compaction_test` demonstrates such behavior. Value whose ref count temporarily went to 0 is readable again because the snapshot was created before the compaction.

More on rocksdb and merge operators: https://github.com/EighteenZi/rocksdb_wiki/blob/master/Merge-Operator-Implementation.md#compaction

Please review, and if this approach sounds good, I'll implement it

Reply:
I mostly understand the idea here, but I definitely don't understand RocksDB well enough to judge whether this is truly safe or not. I don't claim that its unsafe, but it feels very risky to have super safety critical code like this depend on such arcane details of rocksdb.

To me, it seems like it is much better to proceed with the transactional plan, which anybody with a basic understanding of databases can understand / believe-in-the-correctness of. Is there still a big reason to avoid transactions?

Reply:
so I did a local experiment. Rocksdb transactions do support merge operators and they will fail whenever there's a concurrent merge for the same key. 

So remaining disadvantages of using transactions instead of the suggested approach are:
* timeline: we'd need to finish pruning first, then switch from `delete_range` to point-deletes inside of pruner, and only then enable transactions for authority rocksdb instance
* potential lock contention: we're using optimistic transactions, so for any hot object value(trending NFT, etc.), we'd have to keep retrying. 
The main benefit of the current approach is that the plain merge operator never fails, it just writes increments and decrements. And snapshot "hack" just making sure that supporting merges are not squashed prematurely. Although managing snapshot lifetime might be challenging

Reply:
@phoenix-o - thanks for coming up with a pretty neat solution to the problem! I however find the idea of using snapshots which controls compaction behavior to ensure system correctness a little scary tbh (it'd be fine it system correctness was not at stake but we were trying to optimize for performance). I could be wrong but I think RocksDB allows for running compactions which ignore snapshots and my worry is when an operator triggers a manual compaction (with ignore snapshot set to true) through ldb tool, it might run into the race you described above. Transactions are easier to reason about, and might even allow us to get rid of merge operators (key lookup might be expensive with merge as multiple operand need to be squashed at read time, but like you said transaction may require retries). I think if transaction approach is much worse in terms of performance, then perhaps we can do a more deep dive and convince ourselves that snapshot based approach could work.

Reply:
@phoenix-o - thanks for coming up with a pretty neat solution to the problem! I however find the idea of using snapshots which controls compaction behavior to ensure system correctness a little scary tbh (it'd be fine it system correctness was not at stake but we were trying to optimize for performance). I could be wrong but I think RocksDB allows for running compactions which ignore snapshots and my worry is when an operator triggers a manual compaction (with ignore snapshot set to true) through ldb tool, it might run into the race you described above. Transactions are easier to reason about, and might even allow us to get rid of merge operators (key lookup might be expensive with merge as multiple operand need to be squashed at read time, but like you said transaction may require retries). I think if transaction approach is much worse in terms of performance, then perhaps we can do a more deep dive and convince ourselves that snapshot based approach could work.